### PR TITLE
clientv3: connection retry and customizable endpoint selection

### PR DESCRIPTION
**Pull request by heyitsanthony**

### CODE REVIEW

File: clientv3/client.go [hunk 1]
```diff
@@ -57,37 +70,80 @@ func NewFromURL(url string) (*Client, error) {
...
+	defer c.mu.Unlock()
+	errs = c.errors
+	c.errors = nil
+	return errs
+}
+
+// Dial establishes a connection for a given endpoint using the client's config
+func (c *Client) Dial(endpoint string) (*grpc.ClientConn, error) {
+	// TODO: enable grpc.WithTransportCredentials(creds)
+	return grpc.Dial(endpoint, grpc.WithInsecure())
```
Comment (xiang90) [inline clientv3/client.go]: should we also call endpointDialer here?

**Comment by xiang90** [comment]
LGTM

File: clientv3/client.go (hunk 1 above)
Comment (heyitsanthony) [inline clientv3/client.go]: `Client.Dial()` is there so the retry dialer will be able to establish connections with the right creds. (cf. `dialEndpointList`)

Comment (xiang90) [inline clientv3/client.go]: ah. makes sense.

Comment (xiang90) [inline clientv3/client.go]: I just feel we should have an initial dialer that can also follow some initial policy. In another word, I feel the initial dial policy should be same as the retryDial policy in most the cases.

Comment (heyitsanthony) [inline clientv3/client.go]: It's already using RetryDialer to initialize; it should be trivial to add a separate InitDialer if we need it.

**Pull request merged**